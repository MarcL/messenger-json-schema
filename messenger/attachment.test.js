const Ajv = require('ajv');
const ajv = new Ajv({allErrors: true});
const attachmentSchema = require('./attachment');

describe('Attachment message', () => {
    const validImageUri = 'https://www.test.com/test-image.jpg';
    let validate;
    
    beforeEach(() => {
        validate = ajv.compile(attachmentSchema);
    })

    test('Valid message with url', () => {
        const givenMessage = {
            attachment: {
                type: 'image',
                payload: {
                    url: validImageUri,
                    is_reusable: true
                }
            }
        };

        validate(givenMessage);
        
        expect(validate.errors).toBeNull();
    });

    test('Valid message with attachment_id', () => {
        const givenMessage = {
            attachment: {
                type: 'image',
                payload: {
                    attachment_id: '123456789'
                }
            }
        };

        validate(givenMessage);
        
        expect(validate.errors).toBeNull();
    });

    describe('Invalid message', () => {
        const invalidScenarios = [
            {
                testMessage: 'Message has no attachment property',
                givenMessage: {}
            },
            {
                testMessage: 'Message has no attachment.type property',
                givenMessage: {
                    attachment: {}
                }
            },
            {
                testMessage: 'Message has no attachment.payload property',
                givenMessage: {
                    attachment: {
                        type: 'image'
                    }
                }
            },
            {
                testMessage: 'Message has no attachment.payload.url property',
                givenMessage: {
                    attachment: {
                        type: 'image',
                        payload: {}
                    }
                }
            },
            {
                testMessage: 'Message has invalid attachment.payload.url property',
                givenMessage: {
                    attachment: {
                        type: 'image',
                        payload: {
                            url: 'not-a-uri'
                        }
                    }
                }
            },
            {
                testMessage: 'Message has invalid attachment.type property',
                givenMessage: {
                    attachment: {
                        type: 'invalid attachment type',
                        payload: {
                            url: validImageUri
                        }
                    }
                }
            },
            {
                testMessage: 'Message has invalid attachment.payload.is_reusable property type',
                givenMessage: {
                    attachment: {
                        type: 'image',
                        payload: {
                            url: validImageUri,
                            is_reusable: 'not a boolean'
                        }
                    }
                }
            },
        ];

        invalidScenarios.forEach(scenario => {
            const {testMessage, givenMessage} = scenario;

            test(testMessage, () => {        
                validate(givenMessage);

                expect(validate.errors).toMatchSnapshot();
            });
        });
    });
});
