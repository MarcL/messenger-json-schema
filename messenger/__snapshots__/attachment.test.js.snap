// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Attachment message Invalid message Message has invalid attachment.payload.is_reusable property type 1`] = `
Array [
  Object {
    "dataPath": ".attachment.payload.is_reusable",
    "keyword": "type",
    "message": "should be boolean",
    "params": Object {
      "type": "boolean",
    },
    "schemaPath": "#/properties/attachment/properties/payload/properties/is_reusable/type",
  },
]
`;

exports[`Attachment message Invalid message Message has invalid attachment.payload.url property 1`] = `
Array [
  Object {
    "dataPath": ".attachment.payload.url",
    "keyword": "format",
    "message": "should match format \\"uri\\"",
    "params": Object {
      "format": "uri",
    },
    "schemaPath": "#/properties/attachment/properties/payload/properties/url/format",
  },
]
`;

exports[`Attachment message Invalid message Message has invalid attachment.type property 1`] = `
Array [
  Object {
    "dataPath": ".attachment.type",
    "keyword": "enum",
    "message": "should be equal to one of the allowed values",
    "params": Object {
      "allowedValues": Array [
        "audio",
        "video",
        "image",
        "file",
        "template",
      ],
    },
    "schemaPath": "#/properties/attachment/properties/type/enum",
  },
]
`;

exports[`Attachment message Invalid message Message has no attachment property 1`] = `
Array [
  Object {
    "dataPath": "",
    "keyword": "required",
    "message": "should have required property 'attachment'",
    "params": Object {
      "missingProperty": "attachment",
    },
    "schemaPath": "#/required",
  },
]
`;

exports[`Attachment message Invalid message Message has no attachment.payload property 1`] = `
Array [
  Object {
    "dataPath": ".attachment",
    "keyword": "required",
    "message": "should have required property 'payload'",
    "params": Object {
      "missingProperty": "payload",
    },
    "schemaPath": "#/properties/attachment/required",
  },
]
`;

exports[`Attachment message Invalid message Message has no attachment.payload.url property 1`] = `
Array [
  Object {
    "dataPath": ".attachment.payload",
    "keyword": "required",
    "message": "should have required property 'url'",
    "params": Object {
      "missingProperty": "url",
    },
    "schemaPath": "#/properties/attachment/properties/payload/oneOf/0/required",
  },
  Object {
    "dataPath": ".attachment.payload",
    "keyword": "required",
    "message": "should have required property 'attachment_id'",
    "params": Object {
      "missingProperty": "attachment_id",
    },
    "schemaPath": "#/properties/attachment/properties/payload/oneOf/1/required",
  },
  Object {
    "dataPath": ".attachment.payload",
    "keyword": "oneOf",
    "message": "should match exactly one schema in oneOf",
    "params": Object {
      "passingSchemas": null,
    },
    "schemaPath": "#/properties/attachment/properties/payload/oneOf",
  },
]
`;

exports[`Attachment message Invalid message Message has no attachment.type property 1`] = `
Array [
  Object {
    "dataPath": ".attachment",
    "keyword": "required",
    "message": "should have required property 'type'",
    "params": Object {
      "missingProperty": "type",
    },
    "schemaPath": "#/properties/attachment/required",
  },
  Object {
    "dataPath": ".attachment",
    "keyword": "required",
    "message": "should have required property 'payload'",
    "params": Object {
      "missingProperty": "payload",
    },
    "schemaPath": "#/properties/attachment/required",
  },
]
`;
